packages/babel-helper-modules/src/rewrite-live-references.js

```diff
@@ -0,0 +1,335 @@
+    if (rootBinding !== localBinding) return;
+
+    const importData = imported.get(localName);
+    if (importData) {
+      const ref = buildImportReference(importData);
+
+      if (path.parentPath.isCallExpression({ callee: path.node })) {
+        path.replaceWith(t.sequenceExpression([t.numericLiteral(0), ref]));
+      } else if (path.isJSXIdentifier() && t.isMemberExpression(ref)) {
+        const { object, property } = ref;
+        path.replaceWith(
+          t.JSXMemberExpression(
+            t.JSXIdentifier(object.name),
+            t.JSXIdentifier(property.name),
+          ),
+        );
+      } else {
+        path.replaceWith(ref);
+      }
+
+      requeueInParent(path);
+    }
+  },
+
+  UpdateExpression: {
+    exit(path) {
+      const { scope, imported, exported } = this;
+
+      const arg = path.get("argument");
+      if (!arg.isIdentifier()) return;
+      const localName = arg.node.name;
+
+      if (!imported.has(localName) && !exported.has(localName)) {
+        return;
+      }
+
+      // redeclared in this scope
+      if (scope.getBinding(localName) !== path.scope.getBinding(localName)) {
+        return;
+      }
+
+      const exportedNames = exported.get(localName) || [];
+
+      if (exportedNames.length > 0 || imported.has(localName)) {
+        if (
+          path.node.prefix ||
+          (path.parentPath.isExpressionStatement() &&
+            !path.isCompletionRecord())
+        ) {
+          // ++i => (i += 1);
+          path.replaceWith(
+            t.assignmentExpression("+=", arg.node, t.numericLiteral(1)),
+          );
+        } else {
+          const varName = path.scope.generateDeclaredUidIdentifier("old");
+
+          const assignment = t.binaryExpression(
+            path.node.operator.slice(0, 1),
+            varName,
+            t.numericLiteral(1),
+          );
+
+          // i++ => (_tmp = i, i = _tmp + 1, _tmp)
+          path.replaceWith(
+            t.sequenceExpression([
+              t.assignmentExpression("=", varName, arg.node),
+              t.assignmentExpression("=", arg.node, assignment),
+              varName,
+            ]),
+          );
+        }
+      }
+    },
+  },
+
+  AssignmentExpression: {
+    exit(path) {
+      const { scope, seen, imported, exported, requeueInParent } = this;
+
+      if (seen.has(path.node)) return;
+      seen.add(path.node);
+
+      const left = path.get("left");
+      if (left.isIdentifier()) {
+        // Simple update-assign foo += 1; export { foo };
+        // =>   exports.foo =  (foo += 1);
+        const localName = left.node.name;
+
+        // redeclared in this scope
+        if (scope.getBinding(localName) !== path.scope.getBinding(localName)) {
+          return;
+        }
+
+        const exportedNames = exported.get(localName) || [];
+        if (exportedNames.length > 0 || imported.has(localName)) {
+          const assignment = path.node;
+
+          if (imported.has(localName)) {
```

Comment:
This should throw (as well as other code paths that try to modify imports at runtime).

Probably only fixable with a spec mode.

Reply:
Could you clarify? This logic does throw, it just injects logic that throws at runtime (`buildImportThrow`) instead of compile time, since that's how it'd work in a real environment

Reply:
Oh, I missed the `buildImportThrow` near the end; had looked inside the inner if instead